File: opzoo/FocalLoss/FocalLoss.py
import mobula
from mobula.const import req
import numpy as np


@mobula.op.register
class FocalLoss:
    def __init__(self, alpha=0.25, gamma=2):
        self.alpha = alpha
        self.gamma = gamma

    def forward(self, logits, targets):
        if self.req[0] == req.null:
            return
        out = self.y
        out_size = np.prod(out.size()) if callable(out.size) else out.size
        alpha = self.alpha
        gamma = self.gamma
        if self.req[0] == req.add:
            out_temp = self.F.zeros_like(out)
            mobula.func.focal_loss_forward(out_size=out_size, alpha=alpha, gamma=gamma, logits=logits, targets=targets,
                                           outputs=out_temp)
            self.y[:] += out_temp
        else:
            self.y[:] = 0
            mobula.func.focal_loss_forward(out_size=out_size, alpha=alpha, gamma=gamma, logits=logits, targets=targets,
                                           outputs=self.y)

    def backward(self, dy):
        assert self.req[1] == "null"
        alpha = self.alpha
        gamma = self.gamma
        logits = self.X[0]
        targets = self.X[1]
        out_size = np.prod(targets.size()) if callable(
            targets.size) else targets.size
        if self.req[0] == req.add:
            out_temp = self.F.zeros_like(self.dX[0])
            mobula.func.focal_loss_forward(out_size=out_size, alpha=alpha, gamma=gamma, logits=logits, targets=targets,
                                           outputs=out_temp)
            self.dX[0] += out_temp
        else:
            self.dX[0][:] = 0
            mobula.func.focal_loss_backward(out_size=out_size, alpha=alpha, gamma=gamma, logits=logits, targets=targets,
                                            outputs=self.dX[0])
        self.dX[0][:] = self.dX[0][:] * dy

    def infer_shape(self, in_shape):
        assert len(in_shape) == 2
        assert in_shape[0] == in_shape[1]
        return in_shape, [in_shape[0]]

File: opzoo/FocalLoss/FocalLoss.cpp
/*
 * author: kohill
 * github: https://github.com/kohillyang
 */
#include "mobula_op.h"

#if USING_CUDA
#include <cuda.h>
#else
#include <algorithm>
#include <cmath>
using std::exp;
using std::log;
using std::max;
using std::pow;
#endif  // USING_CUDA

namespace mobula {

template <typename T>
MOBULA_DEVICE inline T sigmoid(T x) {
  T max_val = max(static_cast<T>(0), -1 * x);
  T v0 = exp(-max_val);
  return v0 / (v0 + exp(-x - max_val));
}

template <typename T>
MOBULA_DEVICE inline T log_sigmoid(T x) {
  T max_val = max(static_cast<T>(0), -1 * x);
  return -1 * max_val - log(exp(-max_val) + exp(-x - max_val));
}

template <typename T>
MOBULA_KERNEL focal_loss_forward_kernel(const int out_size, T alpha, T gamma,
                                        const T* logits, const T* targets,
                                        T* outputs) {
  parfor(out_size, [&](int index) {
    T y = targets[index];
    T x = logits[index];
    T sigmoid_x = sigmoid(x);
    T sigmoid_neg_x = 1 - sigmoid_x;
    T output = alpha * y * pow(sigmoid_neg_x, gamma) * log_sigmoid(x);
    output += (1 - alpha) * (1 - y) * log_sigmoid(-x) * pow(sigmoid_x, gamma);
    output *= -1;
    outputs[index] = output;
  });
}  // focal_loss_forward_kernel

template <typename T>
MOBULA_KERNEL focal_loss_backward_kernel(const int out_size, T alpha, T gamma,
                                         const T* logits, const T* targets,
                                         T* outputs) {
  parfor(out_size, [&](int index) {
    T y = targets[index];
    T x = logits[index];
    T sigmoid_x = sigmoid(x);
    T sigmoid_neg_x = 1 - sigmoid_x;
    T output = (alpha - 1 - alpha * y) * pow(sigmoid_x, 1 + gamma);
    output += alpha * y * pow(sigmoid_neg_x, gamma + 1);
    output += (alpha - 1) * gamma * (y - 1) * sigmoid_neg_x *
              pow(sigmoid_x, gamma) * log_sigmoid(-x);
    output -= alpha * gamma * sigmoid_x * y * pow(sigmoid_neg_x, gamma) *
              log_sigmoid(x);
    output += sigmoid_x * y * pow(sigmoid_x, gamma);
    output *= -1;
    outputs[index] = output;
  });
}  // focal_loss_backward_kernel

}  // namespace mobula
